### ISSUE TITLE
AP => Bluesky: support usernames that begin or end with `_`,`-`, or `~`

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#1005]
We currently convert these characters to `-`, which [ATProto allows in the middle of handles, but not at the beginning or end](https://atproto.com/specs/handle).

Fediverse usernames that begin or end with these characters are rare, but definitely exist. Examples in #982.

Users need to translate handles themselves in order to search for users across the bridge, so I'm trying to keep that translation UX as simple as possible. Barring a more complicated translation process, I don't yet have a good idea for how to support these usernames.

**Comment by qazmlp** [comment]
It's a bit out there, but maybe IDNA2008 could work here (also for internationalisation)? The Bsky names *are* supposed to be domain names, after all, and rendering IDNs at least for supporting TLDs may be something desireable on their end too.

Unicode has a good FAQ with a bunch of links here: <https://www.unicode.org/faq/idn.html>
The easiest option would be to just use an IDNA2008 library, though.

The scheme isn't *pretty*, per se, but iinm it keeps the ascii-part of the handle readable, at least, and is unambiguous.

**Comment by qazmlp** [comment]
Right, but this is something they wouldn't be able to do anyway with any sort of non-trivial mapping, which you'd need to cover more than a very limited set of cases. They'd have to use https://fed.brid.gy/ to look up the account (once the request function is online), and that can do a complex conversion for them easily.

(You could still give out ambiguous mappings on a first-come-first-serve basis, though, of course.)

Bsky also indexes users by display name, so such-translated bridged users would most likely still appear quickly in search there. I can't test this since my display name is also my handle, but the bridged account is clearly indexed and others do show up in search for their display name specifically.

**Comment by h-2** [comment]
My main account on Mastodon is `__h2__@mastodon.social` so I'd be very interested in the problem getting resolved!

On a related note: could I use a regular domain I own as my BSky handle for the bridged Fedi account? I think the documentation only covers the other way around...

**Comment by snarfed** [comment]
> On a related note: could I use a regular domain I own as my BSky handle for the bridged Fedi account? I think the documentation only covers the other way around...

Hopefully eventually! That's https://github.com/snarfed/bridgy-fed/issues/826

**Comment by Tamschi** [comment]
https://fed.brid.gy/docs#troubleshooting still links to #982 for unsupported handles instead of this one.

#826 is now implemented, but if I'm not mistaken that flow requires the account to already be bridged, so it doesn't work for account currently blocked by this issue here.

**Comment by snarfed** [comment]
> https://fed.brid.gy/docs#troubleshooting still links to #982 for unsupported handles instead of this one.

Fixed, thanks for the nudge!
 
> #826 is now implemented, but if I'm not mistaken that flow requires the account to already be bridged, so it doesn't work for account currently blocked by this issue here.

True!

**Comment by Tamschi** [comment]
What do you think about letting users allocate a DID and set up a custom handle without bridging the account yet?
That would work around the issue here quite nicely without making the default handle mapping more complicated.

**Comment by snarfed** [comment]
Definitely possible!

**Comment by tesaguri** [comment]
Would Bluesky accept DIDs with <code>"alsoKnownAs":["at://[handle.invalid](https://atproto.com/specs/handle#additional-non-syntax-restrictions)"]</code>, or `"alsoKnownAs": null`? If so, I think BF could not only let users allocate a DID, but also start bridging their account right away without allocating a handle (of course the Bluesky Social app would display "⚠Invalid Handle" though).

**Comment by snarfed** [comment]
@tesaguri interesting idea! I _think_ that should work, yes.

**Comment by snarfed** [comment]
We now DM people with these usernames when they try to enable the bridge, a la #758.

**Comment by mackuba** [comment]
So, just to make sure this is being tracked, linking to another thread about this: https://github.com/snarfed/bridgy-fed/issues/982#issuecomment-2959771769

> since Mastodon accounts can now assign a custom Bluesky username when bridged, is there a way for a Mastodon account that starts with _ to bridge this way, using a custom username?
>
> good idea! The catch is that they'd need to specify their custom domain handle when they initially bridge their account, instead of afterward like right now. Still doable though!
> 
> maybe we can do this if the user chooses a custom domain on the UI first?
> - user toggles bridging
> - leave alert next to toggle: "not bridging bc [...]. set a custom domain to start bridging"
> - user sets custom domain
> - profit

I would love it if this was added, I've got at least two people from iOSDev on Mastodon who I'd like to follow on Bluesky, who have handles starting with _ (_inside & _DavidSmith) :)

**Comment by snarfed** [comment]
Thanks @mackuba! Definitely understood. This isn't a high priority for us, but I know you're a great engineer, PRs are welcome, I'd love to work with you if you want to try tackling this!

**Comment by mackuba** [comment]
😎

**Comment by mackuba** [comment]
Hmm… so I was thinking it would make sense to just change it so that the "Set Bluesky handle: [field]" section is *always* shown for an account that is not from Bluesky, regardless if it's bridging enabled or not. So you could first set it, and then enable bridging.

But I think it's not as simple… because to set the username, you need to first assign it in your DNS, and to do that you need to have the DID to put there, and to have the DID you need to be bridged first 🙃

So I think the only way would be to let the account enable bridging, create a DID with "invalid handle" or something, and then show a warning that you need to set a username to have it fully working… (like @Tamschi & @tesaguri suggested above).

**Comment by snarfed** [comment]
@mackuba one useful bit here is that we know which fediverse accounts will need this. We could show the Bluesky handle text box for just them, before they're bridged, and require it when they try to turn bridging on. We'd then have the handle ready when we turn on bridging, so we could use it from the beginning.
